app/components/DateInput.tsx

```diff
@@ -18,6 +21,10 @@ export function DateInput({
   max?: Date;
   "data-cy": string;
   required?: boolean;
+  setDatesInputParentState: Dispatch<
+    SetStateAction<{ finalDateInputDate: Date; index: number }[]>
+  >;
+  keyIndex: number;
```

Comment:
I think the API of the DateInput gets a little clunky like this, these two props are very specific for being used in the "DatesInput" and would possibly keep us from reusing them elsewhere.
Personally I'd prefer to have an `onChange?: (newDate: Date | null) => void` prop (or similar), where the parent can then choose what to put in there and trigger from that, depending on its own needs.

Reply:
The problem that I have at the moment is that (as far as I am aware), the app is only using React Hooks for state management (and what we have here isn't suitable for persisting in the DB even temporarily, since we are preparing the POST form for it). 

The way we're doing it now, we're making the child responsible for updating the state of the parent, rather than the parent be responsible for updating the state of its child components.

==

Personally, I would have used a state management library to handle it rather than a series of workarounds (either context, or Hooks as far as what we have now). Without state management in some kind of form, I (currently) cannot think of a way for the parent component to become aware of the changes in the child component (or vice-versa).

As far as your current suggestion goes: let's say that I do the change you suggested & make the child update its value via an onchange callback function (which it basically already does). Currently I see 1 of 2 specific obstacles that I would need to overcome:

1. How would the parent (`DatesInput`) be aware of the changes when generating a new child component (`DateInput`)? OR:
2. How would the child (`DateInput`) update the state that the parent (`DatesInput`) has access to in order to know what to base the new date value off of when it creates a new child component

Without finding a solution to either of these issues, I don't think I can make the change you're requesting at this moment

Reply:
While I get that we could consider the usage of a state or form management library to help with these situations, I don't think it's too bad with hooks in this case.
I'm not sure if I really understand your last two issues here, maybe I could show roughly what I mean?

So let's say we took my suggestion there, the parent component then would render the input like this. It's essentially the same code, just inside the parent.

```tsx
<DateInput
  id="date"
  name="date"
  defaultValue={
    datesInputState.at(i)?.finalDateInputDate ?? undefined
  }
  min={MIN_DATE}
  max={MAX_DATE}
  data-cy="date-input"
  required
  onChange={(newDate) => {
    setDatesInputState((current) =>
        current.map((obj) => {
          if (obj.index === i) {
            return { ...obj, finalDateInputDate: updatedDate };
          }

          return obj;
        })
      );
  }}
/>
```

The child can still be the "owner" of its own state, and only notify the parent on changes.

```tsx
<input
  id={id}
  type="datetime-local"
  value={dateToYearMonthDayHourMinuteString(date)}
  min={min ? dateToYearMonthDayHourMinuteString(min) : undefined}
  max={max ? dateToYearMonthDayHourMinuteString(max) : undefined}
  onChange={(e) => {
    const updatedDate = new Date(e.target.value);
    setDate(updatedDate);
    // parent onChange
    onChange?.(updatedDate);
  }}
  data-cy={dataCy}
  required={required}
/>
```

Reply:
Oh okay, I didn't think about moving the Hook's setState function into the child's onChange() callback as an initializer. That's nice, I'll give it a shot - thanks!

Reply:
It works! :) that is a much cleaner way of updating the state. Thanks again!